File: LeetCode/Explore/TopInterview/Amazon/ArrayString/76_minimum_window_substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
      string ans;
      int counter = 0, idx = 0, minLen = INT_MAX;
      unordered_map<char, int> mp;
      for (auto &c : t) {
        mp[c]++;
      }
      for (int r = 0, l = 0; r < s.size() && l < s.size(); r++) {
        if (mp[s[r]] > 0) {
          counter++;
        }
        mp[s[r]]--;
        if (counter < t.size()) {
          continue;
        }
        while (l < s.size() && mp[s[l]] < 0) {
          mp[s[l]]++;
          l++;
        }
        if (ans.empty() || r - l + 1 < minLen) {
          idx = l;
          minLen = r - l + 1;
          ans = s.substr(idx, minLen);
        }
      }
      return ans;
    }
};

File: LeetCode/Explore/TopInterview/Amazon/ArrayString/76_minimum_window_substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        
        mp = Counter(t)
        
        l = 0
        counter = 0
        ans = ""
        
        for r, c in enumerate(s):
          if mp[c] > 0:
            counter += 1
          
          mp[c] -= 1
          if counter < len(t):
            continue
            
          while l < len(s) and mp[s[l]] < 0:
            mp[s[l]] += 1
            l += 1
            
          if len(ans) == 0 or r - l + 1 < len(ans):
            ans = s[l : r + 1]
            
        return ans
